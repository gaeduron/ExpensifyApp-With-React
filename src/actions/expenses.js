import uuid from 'uuid';

				////////DISPATCH ACTIONS////////
//ADD EXPENSE
export const addExpense = (
	{
		description = '',
		note = '',
		amount = 0,
		createdAt = undefined
	} = {}
) => (
	{
		type: 'ADD_EXPENSE',
		expense: {
			id: uuid(),
			description,
			note,
			amount,
			createdAt
		}
	}
);

//REMOVE EXPENSE
export const removeExpense = ({id = 0} = {}) => ({
	type: 'REMOVE_EXPENSE',
	id
});

//EDIT EXPENSE
export const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates

});
				////////----------------////////
