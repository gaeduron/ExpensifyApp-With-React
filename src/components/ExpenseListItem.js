import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ description, amount, createdAt, id}) => (
	<div>
		<h3>{description}</h3>
		<p>amount: {amount} - date: {createdAt}</p>
		<Link to={`/edit/${id}`}>Edit</Link>
	</div>
);

export default ExpenseListItem;
