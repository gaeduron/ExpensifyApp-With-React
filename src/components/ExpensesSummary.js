import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total.js';
import numeral from 'numeral';

export const ExpensesSummary = (props) => (
	<div className="page-header">
		<div className="content-container">
			<h1 className="page-header__title">
				Viewing <span>{
					props.expenses.length
				}</span> expense{
					props.expenses.length > 1 && 's'
				} totalling <span>{
					numeral(getExpensesTotal(props.expenses) / 100)
						.format('$0,0.00')
				}</span>
			</h1>
			<div className="page-header__actions">
				<Link className="page-header__link" to="/create">
					<div className="google__login">
						<span className="gtext">Add Expense</span>
					</div>
				</Link>
			</div>
		</div>
	</div>
);

const mapStateToProps = (state) => {
	return {
		expenses: selectExpenses(state.expenses, state.filters)
	}
};

export default connect(mapStateToProps)(ExpensesSummary);
