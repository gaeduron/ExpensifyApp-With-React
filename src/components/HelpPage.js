import React from 'react';

const HelpPage = () => (
	<div>
		Sorry, we can't help you :(
	</div>
);

export default HelpPage;
