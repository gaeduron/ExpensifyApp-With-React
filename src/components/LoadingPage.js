import React from 'react';

const LoadingPage = () => (
	<div className="loader">
		<img className="loader__image" src="/images/lava-lamp-loader.gif"/>
	</div>
);

export default LoadingPage;
