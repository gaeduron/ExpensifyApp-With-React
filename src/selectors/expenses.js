import moment from 'moment';

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
	let visibleExpenses = expenses.filter((expense) => {
		let expenseCreationMoment = moment(expense.createdAt);
		let matchText = 
			text === "" ||
			expense.description.toLowerCase().includes(text.toLowerCase()) || 
			expense.note.toLowerCase().includes(text.toLowerCase());
		let matchStartDate = 
			!startDate || expenseCreationMoment.isSameOrAfter(startDate, 'day');
		let matchEndDate = 
			!endDate || expenseCreationMoment.isSameOrBefore(endDate, 'day');
		if (matchText && matchEndDate && matchStartDate) {
			return expense
		}
	})

	if (sortBy == 'date') {
		return visibleExpenses.sort((a, b) => {
			if (a.createdAt > b.createdAt) { return -1 }
			if (a.createdAt < b.createdAt) { return 1 }
			return 0;
		});
	} else {
		return visibleExpenses.sort((a, b) => {
			if (a.amount > b.amount) { return -1 }
			if (a.amount < b.amount) { return 1 }
			return 0;
		});
	}
};

export default getVisibleExpenses;
