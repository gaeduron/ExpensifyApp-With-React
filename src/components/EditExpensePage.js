import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
	onSubmit = (expense) => {
		this.props.editExpense(this.props.expense.id, expense);
		this.props.history.push('/');
	}
	onClick = () => {
		this.props.removeExpense(this.props.expense.id);
		this.props.history.push('/');
	}

	render () {
		return (
			<div>
				<ExpenseForm
					onSubmit={this.onSubmit}
					expense={this.props.expense}
				/>
				<button onClick={this.onClick}>
				Remove
			</button>
		</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => ({
	editExpense: (id, expense) => dispatch(editExpense(id ,expense)),
	removeExpense: (id) => dispatch(removeExpense({ id: id }))
});

const mapStateToProps = (state, props) => {
	return {
		expense: state.expenses.find((expense) => expense.id === props.match.params.id)
	}
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
