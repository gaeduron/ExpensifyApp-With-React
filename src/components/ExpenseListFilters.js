import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
	state = {
		calendarFocused: null
	}

	onDatesChange = ({startDate, endDate}) => {
		this.props.setStartDate(startDate);	
		this.props.setEndDate(endDate);	
	};

	onFocusChange = (calendarFocused) => {
		this.setState(() => ({ calendarFocused }));	
	};

	onTextChange = (e) => {
					switch (e.target.value) {
						case 'date':
							this.props.sortByDate();
							break;
						case 'amount':
							this.props.sortByAmount();
							break;
					}
				};
	
	onSortChange = (e) => {
					this.props.setTextFilter(e.target.value)	
				};

	render() {
		return (
			<div>
				<input type="text" value={this.props.filters.text} onChange={this.onSortChange}/>
				<select
					value={this.props.filters.sortBy}
					onChange={this.onTextChange}>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
					startDate={this.props.filters.startDate}
					endDate={this.props.filters.endDate}
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={() => false}
					showClearDates={true}
				/>
			</div>
		);
	}

}

const mapDispatchToProps = (dispatch) => ({
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

const mapStateToProps = (state) => {
	return {
		filters: state.filters	
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
