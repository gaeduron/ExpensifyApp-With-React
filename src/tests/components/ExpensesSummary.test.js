import { shallow } from 'enzyme';
import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary correctly with 1 expense', () => {
	const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]}/>);
	expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary correctly with multiples expenses', () => {
	const wrapper = shallow(<ExpensesSummary expenses={expenses}/>);
	expect(wrapper).toMatchSnapshot();
});
