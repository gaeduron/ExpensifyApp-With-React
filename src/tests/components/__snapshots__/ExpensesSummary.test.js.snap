// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       expense
       totalling 
      <span>
        $1.95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="page-header__link"
        replace={false}
        to="/create"
      >
        <div
          className="google__login"
        >
          <span
            className="gtext"
          >
            Add Expense
          </span>
        </div>
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary correctly with multiples expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       expense
      s
       totalling 
      <span>
        $1,141.95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="page-header__link"
        replace={false}
        to="/create"
      >
        <div
          className="google__login"
        >
          <span
            className="gtext"
          >
            Add Expense
          </span>
        </div>
      </Link>
    </div>
  </div>
</div>
`;
